### PR TITLE
Add direct search for exact matches of filenames or trailing path seg…

### PR DESCRIPTION
**Pull request by erikrose**
…ments. Fixes bug 1171313.

* Introduce a new type of direct searcher than can return LINE-domain results.
* Add a new filename direct searcher that uses it.
* Switch test harness to use url_for() rather than hard-coding URLs.
* Change direct_result_eq(..., None) to mean "returns a FILE-domain direct result" rather than "doesn't return a direct result".

### CODE REVIEW

File: dxr/testing.py [hunk 1]
```diff
@@ -30,9 +30,14 @@ class TestCase(unittest.TestCase):
...
 
     def client(self):
-        app = make_app(self.config())
-        app.config['TESTING'] = True  # Disable error trapping during requests.
-        return app.test_client()
+        self._app = make_app(self.config())
+        self._app.config['TESTING'] = True  # Disable error trapping during requests.
+        return self._app.test_client()
+
+    def url_for(self, *args, **kwargs):
```
Comment (pelmers) [inline dxr/testing.py]: I like the refactoring here.

File: dxr/plugins/clang/tests/test_direct.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def test_qualified_function_name_sensitive(self):
     def test_qualified_function_name_multiple_matches(self):
         """Multiple matches on fully qualified function name should return
```
Comment (pelmers) [inline dxr/plugins/clang/tests/test_direct.py]: Does it still make sense to say "should return None" in these docstrings?

Comment (erikrose) [inline dxr/plugins/clang/tests/test_direct.py]: Nope! Good catch

**Comment by pelmers** [comment]
r+

**Pull request merged**